dnslol/dnslol.go

```diff
@@ -0,0 +1,310 @@
+		// This shouldn't ever happen, but be defensive in case of a bug :-)
+		return errors.New(
+			"Experiment selector returned zero DNS server addresses from PickServers")
+	}
+
+	// Build the queries for this name for each of the nameservers
+	queries := e.buildQueries(name, servers)
+
+	// Run the built queries, populating the prometheus result stat according to
+	// the results
+	for _, q := range queries {
+		stats.attempts.With(prom.Labels{"server": q.Server}).Add(1)
+		resultLabels := prom.Labels{"server": q.Server}
+		resultType, success, err := e.queryOne(dnsClient, q)
+		// If the result was an error, put the error string in the result label
+		if err != nil {
+			resultLabels["result"] = err.Error()
+		} else if success {
+			// If the result was successful, increment the success stat and put the
+			// resultType in the result label
+			stats.successes.With(prom.Labels{"server": q.Server}).Add(1)
+			resultLabels["result"] = resultType
+		} else {
+			// Otherwise, if the result was not an error but wasn't a success just put
+			// the resultType in the resultLabels. Don't increment the successes stat.
+			resultLabels["result"] = resultType
+		}
+		// TODO(@cpu): This should be a separate function.
+		if e.PrintResults {
+			outcome := "bad"
+			if success {
+				outcome = "ok"
+			}
+			var line strings.Builder
+			fmt.Fprintf(&line, "Server=%s Name=%s QueryType=%s Result=%s",
+				q.Server, q.Name, dns.TypeToString[q.Type], resultType)
+			if err != nil {
+				fmt.Fprintf(&line, " Error=%s", err.Error())
+			}
+			fmt.Fprintf(&line, " Outcome=%s", outcome)
+			log.Printf("%s", line.String())
+		}
+		stats.results.With(resultLabels).Add(1)
+	}
+	return nil
+}
+
+// buildQueries creates queries for the given name, one per server. The types of
+// queries that are built depends on the Experiment's CheckA, CheckAAAA,
+// CheckTXT and CheckCAA settings.
+func (e Experiment) buildQueries(name string, servers []string) []query {
+	// queryPerServer returns a list with one query per server for the given name
+	// and type.
+	queryPerServer := func(name string, typ uint16) []query {
+		var results []query
+		for _, server := range servers {
+			results = append(results, query{
+				Name:   name,
+				Type:   typ,
+				Server: server,
+			})
+		}
+		return results
+	}
+
+	var queries []query
+	if e.CheckA {
+		queries = append(queries,
+			queryPerServer(name, dns.TypeA)...)
+	}
+	if e.CheckAAAA {
+		queries = append(queries,
+			queryPerServer(name, dns.TypeAAAA)...)
+	}
+	if e.CheckTXT {
+		queries = append(queries,
+			queryPerServer(name, dns.TypeTXT)...)
+	}
+	if e.CheckCAA {
+		// We check CAA differently from other domains by splitting the individual
+		// domain labels up and mimicking CAA tree climbing by making a CAA query
+		// for each label for each server.
+		labels := strings.Split(name, ".")
+		for i := 0; i < len(labels); i++ {
+			queries = append(queries,
+				queryPerServer(strings.Join(labels[i:], "."), dns.TypeCAA)...)
```

Comment:
It indeed does log multiple results for each label of the domain. That seemed reasonable when I assumed it was going to be putting those results into a database soon instead and it would be easier to reconstruct the results from multiple labels for one domain.

Ripping it out for now is fine with me :+1: